File: leetcode/50-powx-n.js
// https://leetcode.com/problems/powx-n/ (top-interview-questions)

/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function(x, n) {
  if (n < 0) {
    x = 1 / x;
  }
  
  const helper = (x, n) => {
    if (n === 0) return 1;
    let res = myPow(x, Math.trunc(n/2));
    res *= res;
    return (n % 2) === 0 ? res : res * x;
  }

  return helper(x, Math.abs(n));
};

File: leetcode/50-powx-n.py
# https://leetcode.com/problems/powx-n/ (top-interview-questions)

class Solution:
    def myPow(self, x: float, n: int) -> float:
        x = 1/x if n < 0 else x
        def helper(x, n):
            if x == 0: return 0
            if n == 0: return 1
            if x == 1 or n == 1: return x
            res = helper(x, n//2)
            res *= res
            return res if n % 2 == 0 else x * res

        result = helper(x, abs(n))
        return result